pkg/wasm/imagefetcher.go

```diff
@@ -0,0 +1,238 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+)
+
+const (
+	wasmPluginFileName = "plugin.wasm"
+)
+
+type ImageFetcherOption struct {
+	Username string
+	Password string
+	// TODO(mathetake) Add signature verification stuff.
+}
+
+func (o *ImageFetcherOption) useDefaultKeyChain() bool {
+	return o.Username == "" || o.Password == ""
+}
+
+type ImageFetcher struct {
+	fetchOpt remote.Option
+}
+
+func NewImageFetcher(opt ImageFetcherOption) *ImageFetcher {
+	var fetchOpt remote.Option
+	// TODO(mathetake): have "Anonymous" option?
+	if opt.useDefaultKeyChain() {
+		// Note that defaukt key chain reads the docker config from DOCKER_CONFIG
+		// so must set the envvar when reaching this branch is expected.
+		fetchOpt = remote.WithAuthFromKeychain(authn.DefaultKeychain)
+	} else {
+		fetchOpt = remote.WithAuth(&authn.Basic{Username: opt.Username})
+	}
+	return &ImageFetcher{
+		fetchOpt: fetchOpt,
+	}
+}
+
+// Fetch is the entrypoint after variant detection the image's manifest annotations.
+func (o *ImageFetcher) Fetch(url string) ([]byte, error) {
```

Comment:
@dgn this will be the entry point for your API implementation, so please put your comments here!